homeassistant/components/fritz/sensor.py

```diff
@@ -138,117 +140,128 @@ def _retrieve_link_attenuation_received_state(
     return status.attenuation[1] / 10  # type: ignore[no-any-return]
 
 
-class SensorData(TypedDict, total=False):
-    """Sensor data class."""
+@dataclass
+class FritzSensorEntityDescription(SensorEntityDescription):
+    """Fritz sensor data class."""
 
-    name: str
-    device_class: str | None
-    state_class: str | None
-    unit_of_measurement: str | None
-    icon: str | None
-    state_provider: Callable
-    connection_type: str | None
+    state_provider: Callable | None = None
+    connection_type: str | None = None
```

Comment:
`state_provider` shouldn't be optional. You can use a mixin class for that. Take a look at #56748 to see how I would have done it.

As you're already doing a large refactoring, I would suggest renaming it to `value_fn`. That has been the goto name for it in other integrations so far.

--
`connection_type` is limited to just one string or `None`. That's a good place to use `Literal`. Also done in #56748